designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,284 @@
+{
+  default: "implicit",
+  type: "string",
+}
+```
+
+The proposed change would impact only cases when the provided value is `undefined`:
+
+| Rule Config             | Current `context.options` | Proposed `context.options` |
+| ----------------------- | ------------------------- | -------------------------- |
+| `"error"`               | `[]`                      | `["implicit"]`             |
+| `["error"]`             | `[]`                      | `["implicit"]`             |
+| `["error", undefined]`  | `[undefined]`             | `["implicit"]`             |
+| `["error", null]`       | `[null]`                  | `[null]`                   |
+| `["error", "explicit"]` | `["explicit"]`            | `["explicit"]`             |
+
+### Example: Nested Object Defaults
+
+Using [`class-methods-use-this`](https://github.com/eslint/eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/lib/rules/class-methods-use-this.js)'s schema as an example:
+
+```json5
+[
+  {
+    type: "object",
+    properties: {
+      exceptMethods: {
+        type: "array",
+        items: {
+          type: "string",
+        },
+      },
+      enforceForClassFields: {
+        type: "boolean",
+        default: true,
+      },
+    },
+    additionalProperties: false,
+  },
+]
+```
+
+- The rule's options parsing logic will no longer need to manually default to `{}`:
+
+  ```diff
+  - const config = Object.assign({}, context.options[0]);
+  + const config = context.options[0];
+  ```
+
+- Options default checks, such as `!==` against `default: true` properties, will no longer be necessary:
+
+  ```diff
+  - const enforceForClassFields = config.enforceForClassFields !== false;
+  + const enforceForClassFields = config.enforceForClassFields;
+  ```
+
+- Options whose provided values don't provide a default don't change behavior:
+
+  ```diff
+  const exceptMethods = new Set(config.exceptMethods || []);
+  ```
+
+The proposed behavior would impact cases when the provided value is `undefined` or is an object without an explicit value for `enforceForClassFields`:
+
+| Rule Config                                   | Current `context.options`            | Proposed `context.options`             |
+| --------------------------------------------- | ------------------------------------ | -------------------------------------- |
+| `"error"`                                     | `[]`                                 | `[ { enforceForClassFields: true } ]`  |
+| `["error"]`                                   | `[]`                                 | `[ { enforceForClassFields: true } ]`  |
+| `["error", undefined]`                        | `[undefined]`                        | `[ { enforceForClassFields: true } ]`  |
+| `["error", null]`                             | `[null]`                             | `[null]`                               |
+| `["error", {}]`                               | `[{}]`                               | `[ { enforceForClassFields: true } ]`  |
```

Comment:
In this case, current `context.options` is `[ { enforceForClassFields: true } ]`. This `default` already works.

Reply:
Ah, right! `useDefaults` is already set: https://github.com/eslint/eslint/blob/22a558228ff98f478fa308c9ecde361acc4caf20/lib/shared/ajv.js#L21

This is what I'd missed when first putting the RFC in review - the note that the built-in `useDefaults` is only a shallow application. Adding in.